Add support for 1559 transactions

sunce86 opened the pull request:
Previously, this code was working only with legacy gas price (single f64 value).
Now, it leverages the new gas price estimators with 1559 support. In case the gas estimators have 1559 support, then the simulation and submit of settlements will be done with 1559 gas price, if not, fallbacks to legacy gas price.

One important thing to notice: previously it was assumed that transaction sent with `x` gas price will either bi mined with `x` gas price or it will stay in transaction pool waiting to be mined or speed up. That said, `effective_gas_price` for legacy transactions is always the same as `x`.
Now, if transaction is sent with `x` estimated 1559 gas price, it can be mined with `y` 1559 gas price in case the transaction is mined in a block with different base_fee_per_gas then the block for which `x` estimated 1559 gas price is originally calculated. That said, `effective_gas_price` for 1559 transactions can be different from the `x`.

Todo: update library dependencies.

Related PRs:
https://github.com/gnosis/gp-gas-estimation/pull/14
https://github.com/gnosis/ethcontract-rs/pull/638
https://github.com/gnosis/gp-transaction-retry/pull/5
https://github.com/gnosis/gp-v2-services/issues/1056

sunce86 committed:
support for 1559 type of transactions and gas price

sunce86 committed:
refactor GasPrice and TypedGasPrice

sunce86 committed:
small fix

nlordell reviewed (approved):
Looks good.

sunce86 commented on solver/src/settlement_simulation.rs:
```diff
@@ -75,7 +77,7 @@ pub async fn simulate_settlements(
                 // solver balance for the default ~150M gas limit. Limit to around the
                 // block gas limit (since we can't fit more anyway).
-                .gas(15_000_000.into());
+                .gas(30_000_000.into());
```
@nlordell 
Is this change ok? I assumed mainnet block size.

sunce86 committed:
small fix

bh2smith commented on orderbook/src/fee.rs:
```diff
@@ -168,7 +168,7 @@ impl MinFeeCalculator {
         kind: Option<OrderKind>,
     ) -> Result<U256, PriceEstimationError> {
-        let gas_price = self.gas_estimator.estimate().await?;
+        let gas_price = self.gas_estimator.estimate().await?.estimate();
```
Seems weird to need to `.estimate()` twice... I can only assume these are two different functions, but which is which.

nlordell replied:
Good point. I think the second estimate is to turn the estimated gas price into a legacy `gas_price` value. Maybe something like `to_legacy_gas_price()` might be more descriptive?

sunce86 replied:
Changed to `effective_gas_price()`, since this term is broadly used and known in ethereum protocol and defi projects.

nlordell replied:
`effective_gas_price` sounds good to me.

bh2smith commented on shared/src/arguments.rs:
```diff
@@ -49,6 +49,9 @@ pub struct Arguments {
     )]
     pub gas_estimators: Vec<GasEstimatorType>,
 
+    #[structopt(long, env)]
+    pub gas_estimator_api_key: Vec<String>,
```
Should this be optional? If it is not set, then we could fall back on whatever we were doing before without this api key? Or possibly have a default value that aligns with the default of `gas_estimators` above. 

Actually, now that I think of it, I would propose that `gas_estimators` be upgraded to a vector of tuples `Vec<(GasEstimatorType, ApiKey)>`. So that we don't have to assert their lengths are equal.

vkgnosis replied:
Imo this should be` blocknative_api_key` . If `GasEstimatorType` includes `BlockNative` we will ensure in `main` that the api key is set. I prefer this over the tuple or vec because not every estimator needs an api key.

sunce86 replied:
Changed to `pub blocknative_api_key: Option<String>`

bh2smith reviewed (approved):
Looks like pretty good progress overall, but I would insist that the configuration parameters involving api key and gas estimator types be merged.

bh2smith reviewed (changes requested):
Changing accidental approval for changes requested in approval message.

sunce86 committed:
unit tests fixed

sunce86 commented on shared/src/gas_price_estimation.rs:
```diff
@@ -44,12 +50,32 @@ pub async fn create_priority_estimator(
+            GasEstimatorType::BlockNative => {
+                if !is_mainnet(&network_id) {
+                    return Err(anyhow!("BlockNative only supports mainnet"));
+                }
```
Fixed

sunce86 committed:
gas_estimator_api_key changed to blocknative_api_key

sunce86 committed:
fix serde rename

sunce86 committed:
cargo dependency fix

sunce86 committed:
cargo fixes, archer settlement fix for gas price comparison

nlordell commented on orderbook/src/fee.rs:
```diff
@@ -168,7 +168,7 @@ impl MinFeeCalculator {
         kind: Option<OrderKind>,
     ) -> Result<U256, PriceEstimationError> {
-        let gas_price = self.gas_estimator.estimate().await?;
+        let gas_price = self.gas_estimator.estimate().await?.effective_gas_price();
```
This is a gotcha! We want the effective gas price for fee computations and not gas caps.

sunce86 committed:
Merge remote-tracking branch 'remotes/origin/main' into sunce86/add-support-for-1559

# Conflicts:
#	e2e/Cargo.toml
#	orderbook/src/fee.rs
#	shared/src/gas_price_estimation.rs
#	solver/src/settlement_submission/retry.rs

sunce86 committed:
reverted

sunce86 committed:
dependencies updated

sunce86 committed:
clippy fix

sunce86 committed:
Merge branch 'main' into sunce86/add-support-for-1559

nlordell committed:
Fix CI for EIP-1559 (#1244)

This PR fixes CI for EIP-1559 compatibility. Unfortunately, Ganache doesn't support EIP-1559 (its missing certain methods and JSON fields in its RPC implementation). There is a Ganache `7.0.0-alpha.1` but it doesn't mine reverted transactions, making it very hard to debug CI failures. For this reason we settled with `hardhat node` for starting up a development node for testing and seems to work with a minor change (specifically, we manually set a gas price of 0 for transactions - this needed because `eth_gasPrice` returns 1 GWEI, and the `gasPrice` setting in the Hardhat configuration does not seem to have an effect on this).

### Test Plan

CI.

sunce86 committed:
Merge branch 'main' into sunce86/add-support-for-1559

sunce86 committed:
Cargo.lock

vkgnosis commented on shared/Cargo.toml:
```diff
@@ -9,15 +9,17 @@ license = "MIT OR Apache-2.0"
 anyhow = "1.0"
 assert_approx_eq = "1.1"
 async-trait = "0.1"
+async-recursion = "0.3"
```
Unused dependency.

sunce86 replied:
Removed.

vkgnosis reviewed (approved):
LGTM.

sunce86 committed:
Merge remote-tracking branch 'remotes/origin/main' into sunce86/add-support-for-1559

# Conflicts:
#	solver/src/driver.rs
#	solver/src/settlement_simulation.rs

sunce86 committed:
removed commented line
